src/rpc/rewards.cpp

```diff
@@ -0,0 +1,762 @@
+            if (pendingTxns.size() == 0) {
+                LogPrint(BCLog::REWARDS, "Failed to generate any transactions for asset '%s'!\n",
+                    asset_name.c_str());
+
+                throw JSONRPCError(RPC_MISC_ERROR, std::string("Failed to distribute reward"));
+            }
+
+            //  Walk through the entire set of transactions to find out if the fees can be covered
+            CAmount totalFees = 0;
+            CAmount totalAmount = 0;
+            for (auto const & pendingTxn : pendingTxns) {
+                totalFees += pendingTxn.fee;
+                totalAmount += pendingTxn.totalAmt;
+            }
+
+            CAmount curBalance = walletPtr->GetBalance();
+            if (curBalance < totalAmount + totalFees) {
+                //  Not Sufficient Funds
+                std::string strError = strprintf("Insufficient funds (%s) to cover reward (%s) as well as fees (%s)!",
+                    FormatMoney(curBalance), FormatMoney(totalAmount), FormatMoney(totalFees));
+                LogPrint(BCLog::REWARDS, "Error: %s\n", strError.c_str());
+
+                responseObj.push_back(Pair("error_nsf", strError));
+            }
+            else {
+                //
+                //  Sufficient Funds... Proceed with transaction commits
+                //
+                bool errorsOccurred = false;
+                UniValue batchResults(UniValue::VARR);
+
+                //  Attempt to commit all generated transactions
+                for (auto & pendingTxn : pendingTxns) {
+                    CValidationState state;
+
+                    if (!walletPtr->CommitTransaction(*pendingTxn.ptr.get(), *pendingTxn.reserveKey.get(), g_connman.get(), state)) {
+                        LogPrint(BCLog::REWARDS, "Error: The transaction was rejected! Reason given: %s\n", state.GetRejectReason());
+
+                        pendingTxn.result->push_back(Pair("error_txn_rejected", state.GetRejectReason()));
+
+                        errorsOccurred = true;
+                    }
+                    else {
+                        //  Remove all successfully processed payments
+                        pendingTxn.payments.clear();
+                    }
+
+                    batchResults.push_back(*pendingTxn.result.get());
+
+                    pendingTxn.payments.clear();
+                }
+
+                if (errorsOccurred) {
+                    responseObj.push_back(Pair("error_rejects", "One or more transactions were rejected"));
+                }
+
+                responseObj.push_back(Pair("batch_results", batchResults));
+            }
+
+            return responseObj;
+        }
+    }
+    else {
+        LogPrint(BCLog::REWARDS, "Failed to retrieve ownership snapshot for '%s' at height %d!\n",
+            asset_name.c_str(), snapshot_height);                
+    }
+
+    throw JSONRPCError(RPC_MISC_ERROR, std::string("Failed to distribute reward"));
```

Comment:
This generic fallthrough message doesn't help the caller much if e.g. we failed to retrieve the ownership snapshot.

Reply:
Should be fixed with f825dcda5198fee39f80ff3b101a44bdcdd4a424